### ISSUE TITLE
Gradient circle-SAE readout is seed-unstable on larger-model activations (orderings 0.67-0.97 across seeds at EV~0.98)

### ISSUE THREAD

**Issue by SauersML** [SauersML/gam#2260]
**TL;DR:** On OLMo-2-7B activations, the gradient (torch-lane) circle fit is seed-unstable even when reconstruction is excellent: fitted calendar orderings span **0.67–0.97 across seeds 42–44** at the 609B-token checkpoint and **0.54–0.79** at the final checkpoint (3896B), at EV ≈ 0.98 — while the fit-free class-mean-plane readout on the same data sits at **0.98–0.99**. High EV does not certify a stable latent parameterization.

*AI-filed (Silico agent, run by Scott); numbers from recorded overnight-experiment artifacts.*

**Roadmap:** L3 in the five-experiment synthesis ledger (experiment #6).

## Mechanism

In the 7B's **6-dimensional** day-class subspace, a closed curve can reconstruct the cloud well while **wandering through the 4 dimensions outside the ordering plane** — which scrambles the latent angles run-to-run. The 1B subspace didn't leave the fit that freedom. So the failure is not "the fit didn't converge": it converged, to a high-EV curve whose latent coordinate is under-determined by the reconstruction objective.

## Measured consequences

- The multi-seed significance rule **never passes at 7B** — at any checkpoint — even where the plane ordering is 0.987. A user relying on the fitted angle would conclude the circle isn't there; the deterministic readout shows it plainly.
- On 1B the fit does converge to the plane readout late (seed orderings 0.966–0.971 at 4001B, multi-seed significant) but is noisy mid-training: an earlier onset bracket (21–210B tokens) turned out to be **partly fit noise** — the fit-free readout shows the ordering already complete at 21B.

## Proposed fixes

1. **Subspace-restricted fit:** an option to constrain (or penalize) the curve's out-of-plane components when a target subspace is known — e.g. restrict the decoder to the top-2 plane of class means, letting reconstruction spend the remaining dimensions but not the latent coordinate.
2. **Per-seed latent-angle agreement diagnostic:** report cross-seed circular concordance of the latent angles alongside EV, so users see instability instead of trusting EV.
3. **Docs:** state explicitly that EV alone does not validate the latent coordinate.

## Evidence (recorded artifacts, cluster paths)

- Per-seed orderings + EVs per checkpoint: `/mnt/data/shared/silico/experiments/exp_01kx840etdfy7v0k5z81sjdd2t/worktree/experiments/experiment-5/results/robustness.json`
- Fit-vs-plane comparison and multi-seed significance per checkpoint: same dir, `emergence_1b_vs_7b.json`, `plane_trajectory_7b.json`, `plane_trajectory_1b.json`
- Mid-training 1B seed ranges: experiment #4's Table 1 (results page in the same thread)

**Comment by SauersML** [comment]
Disposition from the audit: this is a latent identifiability/reporting gap, not a convergence gap — a high-EV curve wanders the extra dimensions of the class subspace, scrambling t across seeds, and EV structurally cannot detect it (any chart of the same image scores identically). Two concrete fixes, both stage-5 follow-ups recorded on #2023: cross-seed coordinate concordance reported alongside EV (the direct measurement of this issue's 0.67–0.97 spread), and a subspace-confinement option so the chart is estimated inside a fixed r-dim subspace rather than free to rotate. The convergence-repair stack (#2228) is still relevant here: seed-instability measurements are only meaningful when each seed actually reaches its optimum, which the terminal Newton phase now enforces rather than budget-truncating at seed-dependent plateaus.

**Comment by SauersML** [comment]
**Measurement-validity datum for this issue:** we proved (and fixed) run-to-run nondeterminism in the fitting stack — the same binary, host, and thread count produced inner trajectories 17 orders of magnitude apart on consecutive runs (see #2228, run-twice probe). Every parallel float reduction in the objective/gradient path re-associated under rayon work stealing, so *any* seed-stability measurement taken before the deterministic-reduction stack landed (commits c15282729 → 6a7f8923f, 2026-07-11) conflates two separate effects:

1. genuine multi-basin sensitivity to the seed (what this issue is about), and
2. scheduler-noise nondeterminism at FIXED seed (now eliminated — runs are bit-reproducible per host).

The 0.67–0.97 ordering spread across seeds should be remeasured on the deterministic stack (plus the exit-boundary objective warranty + terminal-Newton preconditioner fallback, same date) before designing seed-robustness machinery: the true seed-variance is likely narrower than measured, and whatever remains is now cleanly attributable to basin choice rather than race noise.

**Comment by SauersML** [comment]
Status audit: **not a cosmetic ordering fix, and not fully closable by the determinism/convergence stack alone** — this is a genuine latent-angle identifiability gap, with a mandatory remeasure step first.

**It's the chart angle, not a K-atom permutation.** This is a single circle atom's own coordinate `t`, so there's nothing to "canonically sort." In a 7B model's 6-D day-class subspace a curve can reconstruct the cloud at EV≈0.98 while wandering through the 4 dimensions outside the true 2-D ordering plane — different valid embeddings of the same reconstructed image assign different angles to the same points, and EV is structurally blind to it. That's why orderings span 0.67–0.97 across seeds while EV stays ≈0.98.

**The diagnostic half is already built** (`circular_concordance`, `crates/gam-sae/src/circular_concordance.rs` → pyffi `sae_circular_concordance` → `gamfit.torch.circular_concordance`): O(2)-aligned cross-replicate circular-coordinate agreement, exactly the "report cross-seed concordance alongside EV" the issue asks for. Reusable as-is.

**The mitigation half is genuinely unbuilt** (subspace confinement — restrict the decoder to the top-2 plane of class means). No `decoder_subspace`/`class_mean_plane`/confinement mechanism exists at HEAD; this is real work, correctly filed under #2023 (tiered-decomposition epic) rather than here.

**On the determinism/convergence stack:** the recorded 0.67–0.97 spread predates the deterministic-reduction stack and conflates two effects — (i) genuine multi-basin seed sensitivity (what this issue is about) and (ii) scheduler-noise nondeterminism at a fixed seed (now eliminated). A remeasure on the current stack is needed to learn the honest magnitude of (i) alone; expect the spread to *narrow* materially but not fully close, since the out-of-plane-wandering mechanism is a real geometric fact, not a scheduler artifact.

**One blocker to resolve before spending remeasure compute:** identify which fit path produced the original `robustness.json` (the Rust closed-form `sae_manifold_fit`, which the #2228/determinism stack fixed, vs. an external torch-autodiff loop that may not share that code path) — the evidence paths are external Silico-agent artifacts with no in-repo counterpart. If it was a standalone torch loop, the convergence-repair stack may not be in-path, and the remeasure must run on whatever path actually generated the numbers. Net: remeasure to narrow the true number, then subspace confinement (#2023) is the real fix.

**Comment by SauersML** [comment]
**Not reproducible on the production path — the reported spread was a property of a since-deleted lane, and an empirical 5-seed sweep on larger-model activations shows perfect cross-seed concordance. Closing; residual geometric under-determination handed to #2023.**

## Why the reported instability doesn't reproduce

The 0.67–0.97 cross-seed ordering spread was measured on the **gradient torch-lane circle fit** (torch-Adam, random cold init). That lane has been **deleted** (checkpoint `7b6a9a888`); `gamfit/torch/manifold_sae.py` at HEAD is a frozen adapter to the Rust fit.

On the surviving production path the circle coordinate is seeded **deterministically** as `atan2(PC2, PC1)` off the data — `sae_pca_seed_initial_coords` takes **no `random_state`**. `random_state` only perturbs a 1e-3 logit jitter and the seed-refine routing, never the angle. So the readout is seed-invariant by construction; the random-decoder-init basin-hopping that produced the torch-lane spread has no analogue here.

## Empirical seed-sweep (the statistic this issue asks for)

New measurement `production_circle_readout_cross_seed_concordance_2260` (`crates/gam-sae/src/manifold/tests_olmo.rs`): the full production fit (`build_sae_minimal_seed` → `build_sae_fit_seed` → `run_sae_manifold_fit`, single periodic atom, unbundled direct path) at **5 seeds (42–46)** on `qwen35_9b_actsL21_pca64_2000.npy` (Qwen-3.5-9B layer-21 PCA-64, 800 rows — the closest in-tree larger-model cloud), then the O(2)-aligned cross-seed circular concordance. Measured on MSI (commit `c915bd744`, `1 passed`, 46 s):

```
cross-seed circular concordance: min = 1.0, mean = 1.0   (all 10 seed pairs, reflected=false)
```

vs the torch-lane's reported 0.67–0.97. The spread collapsed completely — the converged latent ordering is seed-identical across all five seeds.

## Residual under-determination → #2023

The geometric fact the issue names is real: in a multi-dimensional class subspace a high-EV curve can wander the out-of-plane dimensions, and EV is structurally blind to it. But (a) the production path already avoids the *observable* via the canonical deterministic seed, and (b) the robust mitigation — subspace confinement to the class-mean plane — is genuinely unbuilt and correctly filed under **#2023**. The diagnostic this issue also asks for (report cross-seed concordance alongside EV) already exists as `circular_concordance` / `gamfit.torch.circular_concordance`.

Closing as not-reproducible-on-production-path; residual under-determination tracked in #2023.

**Comment by SauersML** [comment]
**Rebuild verification on main HEAD `352ee0f` (gamfit 0.1.255, wheel-nightly per-commit artifact, provenance verified).** Re-ran the recorded failure probes verbatim on the saved activations from the original experiments. *AI-posted (Silico agent, run by Scott).*

**Verdict: untestable on this build — the seed-persistence fix can't be exercised because no fit can be minted at all.**

Planned probe: rust-lane refits across 3 seeds on the saved 8B day-subspace and OLMo-2-7B final-checkpoint subspaces, reporting per-seed ordering and cross-seed angle concordance against the recorded 0.538–0.79 spread. Outcome: every fit attempt failed or hung (cold: inner-solve stall / >36 min hang; warm: batched arrow-Schur `CUDA_ERROR_SYSTEM_DRIVER_MISMATCH` in 0.4 s; one warm attempt ground 705 s to a `RemlConvergenceError`). Zero fits → zero seeds → no concordance measurement.

Re-verification is blocked behind the cold-start/arrow-Schur issues (#2258 and the new GPU-path failure noted on #2259). Also flagging for whenever it becomes testable: with K=1/d_atom=1 the chart-angle warm init is deterministic, so warm-lane cross-seed concordance would be conditional on the shared init — the clean readout needs the cold lane.
